felix/nftables/table.go

```diff
@@ -0,0 +1,1173 @@
+	for {
+		t.onStillAlive()
+		hashes, rules, err := t.attemptToGetHashesAndRulesFromDataplane()
+		if err != nil {
+			countNumSaveErrors.Inc()
+			var stderr string
+			if ee, ok := err.(*exec.ExitError); ok {
+				stderr = string(ee.Stderr)
+			}
+			t.logCxt.WithError(err).WithField("stderr", stderr).Warn("nftables command failed")
+			if retries > 0 {
+				retries--
+				t.timeSleep(retryDelay)
+				retryDelay *= 2
+			} else {
+				t.logCxt.Panic("nftables command failed after retries")
+			}
+			continue
+		}
+
+		return hashes, rules
+	}
+}
+
+// attemptToGetHashesAndRulesFromDataplane reads nftables state and loads it into memory.
+func (t *nftablesTable) attemptToGetHashesAndRulesFromDataplane() (hashes map[string][]string, rules map[string][]*knftables.Rule, err error) {
+	startTime := t.timeNow()
+	defer func() {
+		saveDuration := t.timeNow().Sub(startTime)
+		t.peakNftablesReadTime = t.peakNftablesReadTime * 99 / 100
+		if saveDuration > t.peakNftablesReadTime {
+			t.logCxt.WithField("duration", saveDuration).Debug("Updating nftables peak duration.")
+			t.peakNftablesReadTime = saveDuration
+		}
+	}()
+
+	t.logCxt.Debug("Attmempting to get hashes and rules from nftables")
+
+	hashes = make(map[string][]string)
+	rules = make(map[string][]*knftables.Rule)
+
+	ctx, cancel := context.WithTimeout(context.Background(), t.contextTimeout)
+	defer cancel()
+
+	// Add chains. We need to query this separately, as chains may exist without rules.
+	allChains, err := t.nft.List(ctx, "chain")
+	if err != nil {
+		if isNotFound(err) {
+			err = nil
+			return
+		}
+		return nil, nil, err
+	}
+	for _, chain := range allChains {
+		hashes[chain] = []string{}
+		rules[chain] = []*knftables.Rule{}
+	}
+
+	// List rules and extract the hashes.
+	allRules, err := t.nft.ListRules(ctx, "")
+	if err != nil {
+		if isNotFound(err) {
+			err = nil
+			return
+		}
+		return nil, nil, err
+	}
+
+	for _, rule := range allRules {
+		// Add the rule to the list of rules for the chain.
+		if _, ok := rules[rule.Chain]; !ok {
+			rules[rule.Chain] = []*knftables.Rule{}
+		}
+		rules[rule.Chain] = append(rules[rule.Chain], rule)
+
+		// Initialize the hashes map for this chain if it doesn't exist.
+		if _, ok := hashes[rule.Chain]; !ok {
+			hashes[rule.Chain] = []string{}
+		}
+
+		if rule.Comment != nil {
+			// The rule has a comment, extract the hash.
+			hash := strings.TrimPrefix(strings.Split(*rule.Comment, ";")[0], t.hashCommentPrefix)
+			hashes[rule.Chain] = append(hashes[rule.Chain], hash)
+		} else {
+			// Otherwise, this rule has no hash and may not be ours. We don't expect these in our chains,
+			// but might appear if someone else has modified our table.
+			hashes[rule.Chain] = append(hashes[rule.Chain], "")
+		}
+	}
+	return
+}
+
+// isNotFound helps distinguish errors resulting from the table not yet being programmed.
+func isNotFound(err error) bool {
+	return strings.Contains(err.Error(), "no such table") ||
```

Comment:
this already exists upstream; `knftables.IsNotFound(err)`

Reply:
@danwinship IIRC, I wasn't able to use `knftables.IsNotFound` here because the Fake implementation doesn't use it - I had a mental note to fix that but I think it got lost :sweat_smile: